### PR TITLE
Add missing sqrt to magSpectrum().

### LINKED ISSUE #20100: DC component sqrt is missing in magSpectrums() called in phaseCorrelate().
In this source code [/modules/imgproc/src/phasecorr.cpp](https://github.com/opencv/opencv/blob/master/modules/imgproc/src/phasecorr.cpp),
I think std::sqrt should be applied to the below lines.

[here](https://github.com/opencv/opencv/blob/master/modules/imgproc/src/phasecorr.cpp#L85) `dataDst[0] = dataSrc[0]*dataSrc[0];`
[here](https://github.com/opencv/opencv/blob/master/modules/imgproc/src/phasecorr.cpp#L87) `dataDst[(rows-1)*stepDst] = dataSrc[(rows-1)*stepSrc]*dataSrc[(rows-1)*stepSrc];`
[here](https://github.com/opencv/opencv/blob/master/modules/imgproc/src/phasecorr.cpp#L104) `dataDst[0] = dataSrc[0]*dataSrc[0];`
[here](https://github.com/opencv/opencv/blob/master/modules/imgproc/src/phasecorr.cpp#L106) `dataDst[j1] = dataSrc[j1]*dataSrc[j1];`
[here](https://github.com/opencv/opencv/blob/master/modules/imgproc/src/phasecorr.cpp#L129) `dataDst[0] = dataSrc[0]*dataSrc[0];`
[here](https://github.com/opencv/opencv/blob/master/modules/imgproc/src/phasecorr.cpp#L131) `dataDst[(rows-1)*stepDst] = dataSrc[(rows-1)*stepSrc]*dataSrc[(rows-1)*stepSrc];`
[here](https://github.com/opencv/opencv/blob/master/modules/imgproc/src/phasecorr.cpp#L148) `dataDst[0] = dataSrc[0]*dataSrc[0];`
[here](https://github.com/opencv/opencv/blob/master/modules/imgproc/src/phasecorr.cpp#L150) `dataDst[j1] = dataSrc[j1]*dataSrc[j1];`

### PR DESCRIPTION
**Pull request by yo1990**
resolves #20100

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under Apache 2 License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

Add some std::sqrt() to magSpectrum().
[#20100](https://github.com/opencv/opencv/issues/20100)

```
force_builders=ARMv7,Custom
build_image:Custom=centos:7
buildworker:Custom=linux-1,linux-4,linux-6
```

### CODE REVIEW

**Comment by llschloesser** [comment]
For [phase correlation](https://en.wikipedia.org/wiki/Phase_correlation), the |  | in DFT' (F G* / |F G*|) means absolute value and not magnitude, right? Is `magSpectrums()` the wrong operation to use for phase correlation?

**Comment by llschloesser** [comment]
> For [phase correlation](https://en.wikipedia.org/wiki/Phase_correlation), the | | in DFT' (F G* / |F G*|) means absolute value and not magnitude, right? Is `magSpectrums()` the wrong operation to use for phase correlation?

In my code I replaced:
`magnitude_of_spectrums(cross_power_, cross_power_magnitude_);`
with
`cv::absdiff(cross_power_, cv::Scalar::all(0), cross_power_magnitude_);`
and observed no loss of accuracy/quality. Same good results as before. What am I not understanding?

**Comment by llschloesser** [comment]
> > For [phase correlation](https://en.wikipedia.org/wiki/Phase_correlation), the | | in DFT' (F G* / |F G*|) means absolute value and not magnitude, right? Is `magSpectrums()` the wrong operation to use for phase correlation?
> 
> In my code I replaced:
> `magnitude_of_spectrums(cross_power_, cross_power_magnitude_);`
> with
> `cv::absdiff(cross_power_, cv::Scalar::all(0), cross_power_magnitude_);`
> and observed no loss of accuracy/quality. Same good results as before. What am I not understanding?

In the book Image Processing, Analysis, and Machine Vision, the magnitude function | F(u,v) | is defined as sqrt(R(u,v)^2 + I(u,v)^2). So it seems that the magnitude function on a spectrum is indeed the right operation and is denoted with single bars rather than double?

File: modules/imgproc/src/phasecorr.cpp [hunk 1]
```diff
@@ -82,9 +82,9 @@ static void magSpectrums( InputArray _src, OutputArray _dst)
             {
                 if( k == 1 )
                     dataSrc += cols - 1, dataDst += cols - 1;
-                dataDst[0] = dataSrc[0]*dataSrc[0];
+                dataDst[0] = (float)std::sqrt(dataSrc[0]*dataSrc[0]);
```
Comment (vpisarev) [inline modules/imgproc/src/phasecorr.cpp:85]: please, replace it with more efficient `std::abs(dataSrc[0])`

File: modules/imgproc/src/phasecorr.cpp [hunk 2]
```diff
@@ -82,9 +82,9 @@ static void magSpectrums( InputArray _src, OutputArray _dst)
...
                 if( rows % 2 == 0 )
-                    dataDst[(rows-1)*stepDst] = dataSrc[(rows-1)*stepSrc]*dataSrc[(rows-1)*stepSrc];
+                    dataDst[(rows-1)*stepDst] = (float)std::sqrt(dataSrc[(rows-1)*stepSrc]*dataSrc[(rows-1)*stepSrc]);
```
Comment (vpisarev) [inline modules/imgproc/src/phasecorr.cpp:87]: please, replace it with more efficient `std::abs(...)`

File: modules/imgproc/src/phasecorr.cpp [hunk 3]
```diff
@@ -101,9 +101,9 @@ static void magSpectrums( InputArray _src, OutputArray _dst)
         {
             if( is_1d && cn == 1 )
             {
-                dataDst[0] = dataSrc[0]*dataSrc[0];
+                dataDst[0] = (float)std::sqrt(dataSrc[0]*dataSrc[0]);
```
Comment (vpisarev) [inline modules/imgproc/src/phasecorr.cpp:104]: use std::abs()

File: modules/imgproc/src/phasecorr.cpp [hunk 4]
```diff
@@ -101,9 +101,9 @@ static void magSpectrums( InputArray _src, OutputArray _dst)
...
                 if( cols % 2 == 0 )
-                    dataDst[j1] = dataSrc[j1]*dataSrc[j1];
+                    dataDst[j1] = (float)std::sqrt(dataSrc[j1]*dataSrc[j1]);
```
Comment (vpisarev) [inline modules/imgproc/src/phasecorr.cpp:106]: use std::abs()

File: modules/imgproc/src/phasecorr.cpp [hunk 5]
```diff
@@ -126,9 +126,9 @@ static void magSpectrums( InputArray _src, OutputArray _dst)
             {
                 if( k == 1 )
                     dataSrc += cols - 1, dataDst += cols - 1;
-                dataDst[0] = dataSrc[0]*dataSrc[0];
+                dataDst[0] = (float)std::sqrt(dataSrc[0]*dataSrc[0]);
```
Comment (vpisarev) [inline modules/imgproc/src/phasecorr.cpp:129]: use abs; in the double-precision version do not cast the result to float

File: modules/imgproc/src/phasecorr.cpp [hunk 6]
```diff
@@ -126,9 +126,9 @@ static void magSpectrums( InputArray _src, OutputArray _dst)
...
                 if( rows % 2 == 0 )
-                    dataDst[(rows-1)*stepDst] = dataSrc[(rows-1)*stepSrc]*dataSrc[(rows-1)*stepSrc];
+                    dataDst[(rows-1)*stepDst] = (float)std::sqrt(dataSrc[(rows-1)*stepSrc]*dataSrc[(rows-1)*stepSrc]);
```
Comment (vpisarev) [inline modules/imgproc/src/phasecorr.cpp:131]: use abs; in the double-precision version do not cast the result to float

File: modules/imgproc/src/phasecorr.cpp [hunk 7]
```diff
@@ -145,9 +145,9 @@ static void magSpectrums( InputArray _src, OutputArray _dst)
         {
             if( is_1d && cn == 1 )
             {
-                dataDst[0] = dataSrc[0]*dataSrc[0];
+                dataDst[0] = (float)std::sqrt(dataSrc[0]*dataSrc[0]);
```
Comment (vpisarev) [inline modules/imgproc/src/phasecorr.cpp:148]: use abs; in the double-precision version do not cast the result to float

File: modules/imgproc/src/phasecorr.cpp [hunk 8]
```diff
@@ -145,9 +145,9 @@ static void magSpectrums( InputArray _src, OutputArray _dst)
...
                 if( cols % 2 == 0 )
-                    dataDst[j1] = dataSrc[j1]*dataSrc[j1];
+                    dataDst[j1] = (float)std::sqrt(dataSrc[j1]*dataSrc[j1]);
```
Comment (vpisarev) [inline modules/imgproc/src/phasecorr.cpp:150]: use abs; in the double-precision version do not cast the result to float

**Comment by vpisarev** [comment]
@yo1990, thank you for the contribution. Please, check the necessary checkboxes. At least, you should agree to contribute your code under Apache 2 license

**Review by vpisarev — APPROVED** [review]

**Comment by vpisarev** [comment]
:+1:

**Comment by yo1990** [comment]
Thank you. 
I agree with your corrections; absolute value is more appropriate, not magnitude.

**Pull request merged**